### PR TITLE
for #117: start telemetry with testpilot-metrics

### PR DESCRIPTION
**Pull request by groovecoder**
Would like to get comments from @bakulf and @jonathanKingston ASAP on this ... does this look like a sane way to add probes?

This commit:

* Adds the `testpilot-metrics` library to the add-on
* Initializes a `sendEvent` function in `index.js`
* Includes 1 example of using the method to send the `container-sort-tabs` ping

If this looks good, I can start to put all the necessary `sendEvent` calls into the rest of the code-base to instrument everything described in metrics.md.

### CODE REVIEW

**Comment by jonathanKingston** [comment]
@groovecoder I carefully ran through the code and it seems to make sense to me. I can't see any issue with how it triggers or obvious race conditions. Want me to test it out too (I know you said ASAP so wanted to respond as quick as possible)?

**Comment by groovecoder** [comment]
Thanks @jonathanKingston. If you want, you can:

1. Install the Test Pilot add-on to your profile
2. Run the .xpi from this branch
3. Click the sort tabs button
4. Go to about:telemetry, archived pings, and look for testpilot-test ping corresponding to your event

**Comment by jonathanKingston** [comment]
@groovecoder these problems likely will be unrelated however these are issues I saw with the XPI:

- [ ] Sorting doesn't seem to work, pressing the button doesn't sort
- [ ] I created a custom container, pink with a briefcase and the url bar for a custom container had a orange icon
- [ ] Alignment for edit issues that @bakulf noticed
- [ ] Longpress button mouseclick doesn't work
- [ ] Specific container siide menu in browser action doesn't work when a container is created (arrow is still greyed out)
- [ ] Container tab count doesn't work for me:
```
{
  "test": "@testpilot-containers",
  "version": "0.0.1",
  "timestamp": 475,
  "variants": null,
  "payload": {
    "event": "container-sort-tabs",
    "total-container-tabs-count": 0,
    "object": null,
    "category": "interactions",
    "variant": null
  }
}
```

Kinda looks like the container service isn't working or suffered an exception. I can't see anything in the browser log. Will do further testing when I get home.

**Comment by groovecoder** [comment]
Thanks @jonathanKingston - can you also check which Fx version you're using? I'm still using unbranded beta 51 for dev.

**Comment by jonathanKingston** [comment]
@groovecoder same was downloaded a while ago: "51.0b7 (64-bit)" on Ubuntu 16.04.

**Comment by groovecoder** [comment]
@jonathanKingston - are you sure you're on the right branch? 

I'm on Nightly (unbranded) 51.0b14 (64-bit) Mac OS OS 10.11.6 ...

* [x] Sorting tabs button works for me on a fresh window with lots of built-in container tabs opened.
* [x] 'custom' container with pink briefcase looks right in url bar
* [x] Edit Containers panel alignment looks good
* [ ] What long-press? I see containers as soon as I hover over the '+' icon in the tab bar
* [x] Container tabs panel works - there has to be an active tab in the container before you can click thru to it
* [x] Container tab count works in the telemetry ping:
```
{
  "test": "@testpilot-containers",
  "version": "0.0.1",
  "timestamp": 121,
  "variants": null,
  "payload": {
    "event": "container-sort-tabs",
    "total-container-tabs-count": 7,
    "object": null,
    "category": "interactions",
    "variant": null
  }
}
```

File: index.js [hunk 1]
```diff
@@ -39,6 +39,13 @@ const webExtension = require("sdk/webextension");
 const windows = require("sdk/windows");
 const windowUtils = require("sdk/window/utils");
 
+const Metrics = require("./testpilot-metrics");
```
Comment (bakulf) [inline index.js]: alphabetic order.

File: index.js [hunk 2]
```diff
@@ -39,6 +39,13 @@ const webExtension = require("sdk/webextension");
...
+const { sendEvent } = new Metrics({
```
Comment (bakulf) [inline index.js]: I would move this into the ContainerService object in the init()

File: index.js [hunk 3]
```diff
@@ -134,6 +141,16 @@ const ContainerService = {
 
   // utility methods
 
+  _totalContainerTabs() {
+    let totalContainerTabs = 0;
+    for (const userContextId in this._identitiesState) {
+      if (this._identitiesState.hasOwnProperty(userContextId)) {
```
Comment (bakulf) [inline index.js]: remove this line. Just do:

for (const userContextId in this._identitiesState) {
  totalContainerTabs += this._identitiesState[userContextId].openTabs;
}

File: index.js [hunk 4]
```diff
@@ -310,6 +327,10 @@ const ContainerService = {
   },
 
   sortTabs() {
+    sendEvent({
```
Comment (bakulf) [inline index.js]: I don't like this :)
I prefer this:

this._sendTelemetryData('container-sort-tabs');

where _sendTelemetryData is:

function _sendTelemetryData(event) {
  let totalContainerTabsCount = 0;
  for ... identitiesState...

  this._sendEvent({
    event,
    "total-container-tabs-count": totalContainerTabsCount,
  });
}

File: testpilot-metrics.js [hunk 5]
```diff
@@ -0,0 +1,333 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (bakulf) [inline testpilot-metrics.js]: Usually, what I think we should do when we take an external file, is to have a separate commit _just_ with this file. and then in separate commits we start using it.

File: index.js (hunk 4 above)
Comment (bakulf) [inline index.js]: If we use this approach, we can easily call: this._sendTelemetryData in other parts of the code.

Comment (jonathanKingston) [inline index.js]: Assuming we always want to send the tab count this was something I was going to suggest.

**Comment by jonathanKingston** [comment]
@groovecoder yeah:
```
git log -n 1
commit 187df364b0256363a1add748e2fcd17834f79b80
Author: groovecoder <luke.crouch@gmail.com>
Date:   Thu Feb 9 10:59:29 2017 -0600

    for #117: start telemetry with testpilot-metrics
```

However I may have had a stale build in my directory, I deleted the xpi's in my directory and tried to load a new one in about:debugging and got this:
```
Error: File /home/jonathan/projects/testpilot-containers/@testpilot-containers-0.0.1.xpi does not contain a valid install manifest
Stack trace:
loadManifestFromZipReader<@resource://gre/modules/addons/XPIProvider.jsm:1482:11
TaskImpl_run@resource://gre/modules/Task.jsm:319:40
TaskImpl@resource://gre/modules/Task.jsm:280:3
createAsyncFunction/asyncFunction@resource://gre/modules/Task.jsm:254:14
loadManifestFromZipFile<@resource://gre/modules/addons/XPIProvider.jsm:1539:26
TaskImpl_run@resource://gre/modules/Task.jsm:319:40
TaskImpl@resource://gre/modules/Task.jsm:280:3
createAsyncFunction/asyncFunction@resource://gre/modules/Task.jsm:254:14
loadManifestFromFile@resource://gre/modules/addons/XPIProvider.jsm:1549:12
this.XPIProvider.installTemporaryAddon<@resource://gre/modules/addons/XPIProvider.jsm:4031:23
TaskImpl_run@resource://gre/modules/Task.jsm:319:40
TaskImpl@resource://gre/modules/Task.jsm:280:3
createAsyncFunction/asyncFunction@resource://gre/modules/Task.jsm:254:14
AddonManagerInternal.installTemporaryAddon@resource://gre/modules/AddonManager.jsm:2329:12
this.AddonManager.installTemporaryAddon@resource://gre/modules/AddonManager.jsm:3506:12
loadAddonFromFile@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/aboutdebugging/components/addons/controls.js:57:5
invokeGuardedCallback@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10051:12
executeDispatch@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:2828:5
executeDispatchesInOrder@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:2851:5
[16]</executeDispatchesAndRelease@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:2291:5
[16]</executeDispatchesAndReleaseTopLevel@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:2302:10
[124]</forEachAccumulated@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:17830:5
[16]</EventPluginHub.processEventQueue@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:2507:7
runEventQueueInBatch@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10120:3
[62]</ReactEventEmitterMixin.handleTopLevel@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10136:5
handleTopLevelWithoutPath@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10234:5
handleTopLevelImpl@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10214:3
[113]</Mixin.perform@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:17257:13
[53]</ReactDefaultBatchingStrategy.batchedUpdates@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:8849:7
batchedUpdates@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:15333:3
[63]</ReactEventListener.dispatchEvent@resource://gre/modules/commonjs/toolkit/loader.js -> resource://devtools/client/shared/vendor/react.js:10345:7
  controls.js:59:9
```

File: index.js (hunk 4 above)
Comment (bakulf) [inline index.js]: We are not going to have the same payload for each event, but we can have a method like this:

this._sendEvent(event, "totalTab", "foobar", "barfoo");

and sendEvent parses the arguments in order to create the correct payload.

File: index.js (hunk 1 above)
Comment (groovecoder) [inline index.js]: I pulled this down below the others because I tend to do my imports in this structure:

1. Lowest-level/common-library modules
2. 3rd-party library modules
3. local modules
4. define local constants

But I'm happy to do alphabetical if that makes more sense.

**Comment by groovecoder** [comment]
@jonathanKingston try the new 0.0.2 xpi?

File: index.js (hunk 4 above)
Comment (groovecoder) [inline index.js]: I'll try this, but I'm not sure an extra `_sendTelemetryData` utility method will give us any better code re-use than a `this._sendEvent` method gives us.

**Comment by groovecoder** [comment]
I tried a bit to create a `_sendTelemetryPing` helper method, but until we implement more pings I can't quite predict which logic should be in the helper vs. the calling code. So I'm punting on that for now.

But I think I addressed everything else? Look good enough now for me to add a dozen or so more calls to `_sendEvent` - understanding that I'll factor out a `_sendTelemetryPing` along the way if I can.

**Comment by groovecoder** [comment]
Okay @jonathanKingston and @bakulf I need another review here. I added a couple more helper methods to get the `uuid`, `totalContainerTabsCount`, `totalNonContainerTabsCount`, and `shownContainersCount` to add to the `sort-tab` event as the first event we're recording with its full telemetry payload.

Please review ASAP before I make more calls to these methods for the rest of the telemetry probe payloads.

File: index.js [hunk 6]
```diff
@@ -350,6 +401,13 @@ const ContainerService = {
   },
 
   sortTabs() {
+    this._sendEvent({
+      "uuid": this._metricsUUID,
```
Comment (groovecoder) [inline index.js]: This should move up into `_sendEvent`

File: testpilot-metrics.js (hunk 5 above)
Comment (groovecoder) [inline testpilot-metrics.js]: Moved this to a separate commit.

File: index.js (hunk 6 above)
Comment (groovecoder) [inline index.js]: Moved.

**Comment by groovecoder** [comment]
Now that we have base `_sendEvent` and `_sendTelemetryPayload` methods, https://github.com/mozilla/testpilot-containers/pull/163/commits/7edfd275864bcbb26118b8bbbd45ee441cd688a4 shows how they are re-used to send telemetry for `open-tab`.

File: index.js [hunk 7]
```diff
@@ -60,13 +63,16 @@ const ContainerService = {
     if (installation) {
       const object = {
         version: 1,
-        prefs: {}
+        prefs: {},
+        metricsUUID: null
```
Comment (bakulf) [inline index.js]: add an extra ',' at the end

Comment (bakulf) [inline index.js]: metricsUUID: uuid.uuid().toString(),

File: index.js [hunk 8]
```diff
@@ -60,13 +63,16 @@ const ContainerService = {
...
       };
 
       PREFS.forEach(pref => {
         object.prefs[pref[0]] = prefService.get(pref[0]);
       });
 
+      object.metricsUUID = uuid.uuid().toString();
```
Comment (bakulf) [inline index.js]: remove this line and move it in the CTOR of the object.

File: index.js [hunk 9]
```diff
@@ -76,6 +82,15 @@ const ContainerService = {
       prefService.set(pref[0], pref[1]);
     });
 
+    if (ss.storage.savedConfiguration.hasOwnProperty("metricsUUID")) {
+      this._metricsUUID = ss.storage.savedConfiguration.metricsUUID;
+    } else {
+      // The add-on was installed before metricsUUID was added, create one
+      this._metricsUUID = uuid.uuid().toString();
+      ss.storage.savedConfiguration["metricsUUID"] = this._metricsUUID;
```
Comment (bakulf) [inline index.js]: We don't need to support old version of this file. Just consider this 'metricsUUID' as part of version 1.

File: index.js [hunk 10]
```diff
@@ -158,10 +174,53 @@ const ContainerService = {
...
+      type: "sdk",
+      id: self.id,
+      version: self.version
+    }).sendEvent;
+
   },
 
   // utility methods
 
+  _containerTabCount(userContextId) {
```
Comment (bakulf) [inline index.js]: Just to make this code easy to follow, add a comment: "This method returns the total of open and hidden tabs with this userContextId" or something similar...

File: index.js [hunk 11]
```diff
@@ -158,10 +174,53 @@ const ContainerService = {
...
+    let containerTabsCount = 0;
+    containerTabsCount += this._identitiesState[userContextId].openTabs;
+    containerTabsCount += this._identitiesState[userContextId].hiddenTabs.length;
+    return containerTabsCount;
+  },
+
+  _totalContainerTabsCount() {
```
Comment (bakulf) [inline index.js]: "Here we return the number of total container opened tabs..."

File: index.js [hunk 12]
```diff
@@ -158,10 +174,53 @@ const ContainerService = {
...
+    let totalContainerTabsCount = 0;
+    for (const userContextId in this._identitiesState) {
+      totalContainerTabsCount += this._identitiesState[userContextId].openTabs;
+    }
+    return totalContainerTabsCount;
+  },
+
+  _totalNonContainerTabsCount() {
```
Comment (bakulf) [inline index.js]: comment here as well.

File: index.js [hunk 13]
```diff
@@ -158,10 +174,53 @@ const ContainerService = {
...
+    let totalNonContainerTabsCount = 0;
+    for (const tab of tabs) {
+      if (this._getUserContextIdFromTab(tab) === 0) {
+        ++totalNonContainerTabsCount;
+      }
+    }
+    return totalNonContainerTabsCount;
+  },
+
+  _shownContainersCount() {
```
Comment (bakulf) [inline index.js]: A comment here.

File: index.js [hunk 14]
```diff
@@ -504,6 +580,13 @@ const ContainerService = {
         userContextId = args.userContextId;
       }
 
+      this.sendTelemetryPayload({
+        "event": "open-tab",
+        "eventSource": args.source,
+        "userContextId": userContextId,
```
Comment (bakulf) [inline index.js]: userContextId,

File: index.js [hunk 15]
```diff
@@ -357,6 +349,17 @@ const ContainerService = {
     ss.storage.identitiesData = this._identitiesState;
   },
 
+  sendTelemetryPayload(params = {}) {
```
Comment (bakulf) [inline index.js]: can you call this 'args' as all the other methods?

File: index.js [hunk 16]
```diff
@@ -504,6 +580,13 @@ const ContainerService = {
         userContextId = args.userContextId;
```
Comment (bakulf) [inline index.js]: you must add at line 500:

if (!("source" in args)) {
  return Promise.reject(...

File: index.js [hunk 17]
```diff
@@ -290,6 +349,17 @@ const ContainerService = {
...
 
+  sendTelemetryPayload(params = {}) {
+    // when pings come from popup, delete "method" prop
+    delete params.method;
+    let payload = { // eslint-disable-line prefer-const
+      "uuid": this._metricsUUID
+    };
+    Object.assign(payload, params);
+
+    this._sendEvent(payload);
```
Comment (bakulf) [inline index.js]: I wonder if we can have the same payload (or something super super similar) in each telemetry event.
Or move the complexity of creating the payload in 1 single place.

For instance, my initial idea was to have a method like this:

_sendEvent(eventType, userContextId, configuration) {
  let payload = Telemetry.createPayload(eventType, userContextId, configuration);
  ...
}

But we can do this later, if the complexity increases.

Comment (groovecoder) [inline index.js]: Yeah I'll try to consolidate as much complexity into a single place as I can where the payloads are similar enough.

File: index.js (hunk 16 above)
Comment (groovecoder) [inline index.js]: Not sure we have to reject? I'll change to set a default value if `sources` isn't in the `args` in case I'm missing a source of calling code.

**Comment by groovecoder** [comment]
Think I addressed all of @bakulf comments.

**Comment by groovecoder** [comment]
Okay, I've added all the remaining telemetry probes according to [the current metrics.md file](https://github.com/mozilla/testpilot-containers/blob/976925108d6daadb4dc6237cecbd21e8884046f0/docs/metrics.md).

@bakulf @jonathanKingston - last review?
